Make Excluding() work on nested collections if root is a collection

whymatter opened the pull request:
Fixes #1989 #1919

I removed the `!ContainsIndexingQualifiers(selectedPath)` check on this piece of code which was introduced here: https://github.com/fluentassertions/fluentassertions/commit/aab3303a1a227b6c35d66278ed88547dbc58f97a#diff-53539b5c2a4cdb7276ff5e63cbe410a164938d3620ebf08f20797810d22b394c. I am 100% unsure what it does and what the implications are of me removing it. However, there are no failing tests with the condition removed.

```c#
if (!ContainsIndexingQualifiers(selectedPath))  // <-- removed this
{
    currentPath = RemoveIndexQualifiers(currentPath);
}
```

If removing this condition is bad for whatever reason, please tell me which test case I can add to cover this.

## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [ ] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [ ] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

whymatter commented on Src/FluentAssertions/Equivalency/Selection/SelectMemberByPathSelectionRule.cs:
```diff
@@ -7,32 +6,12 @@ namespace FluentAssertions.Equivalency.Selection;
-
-        // If we're part of a collection comparison, the selected path will not include an index,
-        // so we need to remove it from the current node as well.
-        if (!ContainsIndexingQualifiers(selectedPath))
```
This is the part where I am not sure about the implications, since all tests are green

whymatter commented:
Hmm, I don´t understand it. `RemoveIndexQualifiers` does only something if the root is a collection, right? So that the following is true 
```c#
RemoveIndexQualifiers("root[0].Foo[2].Bar") = "root.Foo[2].Bar" // root is collection
RemoveIndexQualifiers("root.Foo[2].Bar") => "root.Foo[2].Bar".  // root is no collection
```
The if condition I removed prevents the first case from happening if I add an exclusion like `x => x.Foo[2].Bar`. Why?
What is a test case for the if condition? I don't get it :(

dennisdoomen commented:
> Hmm, I don´t understand it. `RemoveIndexQualifiers` does only something if the root is a collection, right?

No, it was called when the path contained _any_ `[` and `]`, regardless of the root. This was used in _expressions_, not strings.

So `opt.Excluding(x => x.Foo[2].Bar)` would get normalized to `root.Foo[].Bar`, which is the old way of doing `opt.For(x => x.Foo).Excluding(y => .Bar)`.

whymatter commented:
So is my PR fine then?

dennisdoomen commented:
> So is my PR fine then?

I'm not sure anymore if my memory is serving me well. But the issue was that if we were using `BeEquivalentTo` on a collection, the caller could provide the property path for including or excluding a property without the first index (e.g. `root[0]`). Instead, it would specify the path that assumes an item of the collection. But the actual path of the node in the graph _will_ include `root[0]` or something like that.

whymatter commented:
> Also, if I remove that part, I have about 30 tests failing

I assume you removed the line `currentPath = RemoveIndexQualifiers(currentPath);` as well? I just removed the if condition so that `currentPath = RemoveIndexQualifiers(currentPath);` executes all the time

dennisdoomen commented:
> I assume you removed the line `currentPath = RemoveIndexQualifiers(currentPath);` as well? I just removed the if condition so that `currentPath = RemoveIndexQualifiers(currentPath);` executes all the time

Yeah, I removed the entire `if` statement.

whymatter commented:
> > I assume you removed the line `currentPath = RemoveIndexQualifiers(currentPath);` as well? I just removed the if condition so that `currentPath = RemoveIndexQualifiers(currentPath);` executes all the time
> 
> Yeah, I removed the entire `if` statement.

Okay, so because `currentPath = RemoveIndexQualifiers(currentPath);` is definitely required.
However, I do not understand if and why the `if` surrounding it is required? (Especially because removing the `if` resolves the bug I am working on)

dennisdoomen commented:
> However, I do not understand if and why the `if` surrounding it is required? (Especially because removing the `if` resolves the bug I am working on)

Me neither. So if all the tests succeed, I guess this should be good enough. I can't think of a scenario where we would not want to remove the starting `[0]`

whymatter commented on Src/FluentAssertions/Equivalency/Selection/SelectMemberByPathSelectionRule.cs:
```diff
@@ -7,32 +6,12 @@ namespace FluentAssertions.Equivalency.Selection;
-            currentPath = RemoveIndexQualifiers(currentPath);
-        }
-
+        var currentPath = RemoveIndexQualifiers(currentNode.PathAndName);
```
I like it, however, renamed it to the singular `RemoveRootIndexQualifier`

jnyrup commented on docs/_pages/releases.md:
```diff
@@ -15,6 +15,8 @@ sidebar:
 ### Fixes
 * Improved robustness of several assertions when they're wrapped in an `AssertionScope` - [#2133](https://github.com/fluentassertions/fluentassertions/pull/2133)
 
+* Fixed `.Excluding()`, `.For().Exclude()` not working if root is a collection - [#2101](https://github.com/fluentassertions/fluentassertions/pull/2135)
```
```suggestion
* Fixed `.Excluding()` and `.For().Exclude()` not working if root is a collection - [#2135](https://github.com/fluentassertions/fluentassertions/pull/2135)
```
